src/libra_py/CP2K_methods.py

```diff
@@ -197,55 +199,354 @@ def read_cp2k_tddfpt_log_file( tddfpt_log_file_name, num_ci_states, tolerance ):
-                # 'R-TDDFPT states of multiplicity 1' in the log file
+                # Here we search for the line that contains 'R-TDDFPT states of multiplicity 1'
+                # or 'U-TDDFPT states of multiplicity 1' in the log file
                 r_tddfpt_line = i
 
     excitation_energies = []
-    
+
     # Start from 5 lines after finding the line contaning 'R-TDDFPT states of multiplicity 1'
     # This is because they contain the energies from that line.
     for i in range( r_tddfpt_line+5, len( lines ) ):
         if len( lines[i].split() ) == 0:
             break
         excitation_energies.append( float( lines[i].split()[2] ) )
-        
+
     # Start from 5 lines after finding the line contaning 'Excitation analysis'
     # From that point we have the state numbers with their configurations.
     # So, we append the lines which contain only 'State number' and stop
     # whenever we reach to a blank line.
     state_num_lines = []
     for i in range( exc_anal_line+5, len( lines ) ):
 
-        if len( lines[i].split() ) == 0:
-            state_num_lines.append( i )
-            break
+        if isUKS == 1:
+
+            if len( lines[i].split() ) == 0 or '----' in lines[i]:
+                state_num_lines.append( i )
+                break
+            if len(lines[i].split())==1 or len(lines[i].split())==3:
+                state_num_lines.append(i)
+
+        else:
+
+            if len( lines[i].split() ) == 0 or '----' in lines[i]:
+                state_num_lines.append( i )
+                break
+            if len(lines[i].split())==1:
+                state_num_lines.append(i)
+            elif len(lines[i].split())>1:
+                if lines[i].split()[0].isdigit() and not(lines[i].split()[1].isdigit()):
+                    state_num_lines.append( i )
 
-        if len( lines[i].split() ) == 1:
-            state_num_lines.append( i )
 
     # Setting up the CI-basis list and their coefficients list
-    ci_basis        = []
+    ci_basis = []
     ci_coefficients = []
-    for i in range( num_ci_states ):
-        
+    spin_components = []
+    for i in range( number_ci_states ):
+
         # CI states and their coefficients for each excited state
         tmp_ci_state              = []
         tmp_ci_state_coefficients = []
+        tmp_spin = []
         for j in range( state_num_lines[i]+1, state_num_lines[i+1] ):
 
-            if abs( float( lines[j].split()[2] ) ) > tolerance:
-                tmp_ci_state.append( [ int( lines[j].split()[0] ), int( lines[j].split()[1] ) ]  )
-                tmp_ci_state_coefficients.append( float( lines[j].split()[2] ) )
+            # If we are using the spin-polarized Kohn-Sham basis, then the size of the split line is different
+            # from the size of the split line when using the spin-unpolarized Kohn-Sham basis
+            if isUKS == 1:
+
+                ci_coefficient = float( lines[j].split()[4] )
```

Comment:
as in another place I commented, please use temporary objects to reduce the number of function calls, like  ".split", ".replace", etc.
basically, you want to maximize the reuse of the operations already performed on some intermediate variables